8248877: Document API contract for MetaspaceObj subtypes

iklam opened the pull request:
This PR adds documentation for the (non-virtual) functions that subtypes of MetaspaceObj are required to implement in order for certain templates to work properly. 

Requested by @rose00 (sorry I took a long time ....).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8248877](https://bugs.openjdk.java.net/browse/JDK-8248877): Document API contract for MetaspaceObj subtypes


### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to b0db2f3609ecf1374eb5273199f4072533bda2d6
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4327/head:pull/4327` \
`$ git checkout pull/4327`

Update a local copy of the PR: \
`$ git checkout pull/4327` \
`$ git pull https://git.openjdk.java.net/jdk pull/4327/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4327`

View PR using the GUI difftool: \
`$ git pr show -t 4327`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4327.diff">https://git.openjdk.java.net/jdk/pull/4327.diff</a>

</details>

coleenp commented on src/hotspot/share/memory/allocation.hpp:
```diff
@@ -263,6 +263,24 @@ class ClassLoaderData;
+  //   MetaspaceObj::Type type() const { return <This>Type; }
+  //
+  // The following functions are required by MetadataFactory::free_metadata():
+  //   bool on_stack() { return false; } // DEBUG only
```
I don't think the // DEBUG only comment is correct.  on_stack is used in product mode.  Can you remove this?

iklam replied:
Hmm .... in product code, there's no use of `MetaspaceObj::on_stack()`, but there's use of `Metadata::on_stack()`.

coleenp commented on src/hotspot/share/memory/allocation.hpp:
```diff
@@ -263,6 +263,24 @@ class ClassLoaderData;
+  // Within the sub-hierarchy of Metadata, these are virtuals.
+  // Elsewhere in the hierarchy of MetaspaceObj, type() and/or size() can be static
+  // if constant.
+
```
maybe move this up to be under 275 because it talks about those.

coleenp reviewed (approved):
Looks good with some minor suggestions.

iklam commented:
I cleaned up the comments per Coleen's comments. I also removed the "debug only" comment about on_stack() since in most cases they are inline functions and will be elided by the compiler anyway.

yminqi reviewed (approved):
Looks good.

iklam commented:
Thanks @coleenp and @yminqi for the review.
/integrate
